File: database-driver-api/src/main/java/de/lino/database/utils/Pair.java
package de.lino.database.utils;

import org.jetbrains.annotations.NotNull;

/**
 * A simple, immutable holder for two related values of possibly different types.
 *
 * @param <T>    the type of the first value
 * @param <R>    the type of the second value
 * @param first  the first value
 * @param second the second value
 */
public record Pair<T, R>(@NotNull T first, @NotNull R second) {

}

File: python/database-driver-api/src/database_driver/api/utils/pair.py
"""Mirror of ``de.lino.database.utils.Pair``."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Generic, TypeVar

T = TypeVar("T")
R = TypeVar("R")


@dataclass(frozen=True, slots=True)
class Pair(Generic[T, R]):
    """A simple, immutable holder for two related values of possibly different types.

    Kept as its own type (rather than a plain ``tuple``) so call sites such as
    ``DatabaseRepository.convert`` keep the Java API's named ``first``/``second``
    accessors instead of positional indexing.
    """

    first: T
    second: R
